### PR TITLE
layers: GH544 Allow INVALID cmd buffers to be reset

### PR DESCRIPTION
**Pull request by tobine**
Fix for #544 
BeginCommandBuffer performs an implicit reset of a command buffer.
This allows command buffers in the INVALID state to undergo this
implicit reset.

In order to make sure that an invalid command buffer in the INVALID
state has had EndCommandBuffer called, this commit adds a CMD_END
command that's inserted at the end of a command. If a command buffer
in the INVALID state has had EndCommandBuffer called and it is now
having BeginCommandBuffer called, it's ok to implicitly reset the
command buffer.
buffer's vector of commands.

### CODE REVIEW

**Comment by mark-lunarg** [comment]
Looks good.

**Comment by tobine** [comment]
ping. I just had to add a 2nd CL to prevent what I believe was an invalid error in smoketest due to incorrect comands in a subpass.

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -6251,9 +6254,7 @@ VKAPI_ATTR VkResult VKAPI_CALL EndCommandBuffer(VkCommandBuffer commandBuffer) {
             // https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/pull/516#discussion_r63013756
             skipCall |= insideRenderPass(dev_data, pCB, "vkEndCommandBuffer");
         }
-        if (pCB->state != CB_RECORDING) {
```
Comment (lentinem) [inline layers/core_validation.cpp]: Why was this if removed? It still is true that it must have begun before being ended.

Comment (tobine) [inline layers/core_validation.cpp]: addCmd() call includes that check and more

**Comment by lentinem** [comment]
+1

**Comment by tobine** [comment]
cherry-picked and pushed to master

**Pull request closed without merging**